### PR TITLE
Redfish Client & Redfish OOBM Driver

### PR DESCRIPTION
**Pull request by GabrielBrascher**
## Description

This PR adds support for the OOBM Redfish protocol, implementing a Java client to send HTTP requests to Redfish supported systems.

Implementation overview:
- Redfish Java client: a Java Client for Redfish that makes Redfish actions available to the HA workflow via an OOB driver.
- OOB Redfish driver: a new Out-of-band driver was created for Redfish, allowing to integrate the Redfish Client with the CloudStack Out-of-band management implementation.

Fixes: #3624

**Thanks note:** this implementation had valuable help from our friends at [Yourhosting](https://www.yourhosting.nl/).

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):
Configure OOBM with Redfish driver
image
Check OOBM host tab
image
Test powering On and Off hosts. Host status is properly shown
image

## How Has This Been Tested?

We have been able to test on Dell and Supermicro machines and it works fine:

1. via the CloudStack UI, enable Redfish oob driver for a KVM node with support for Redfish
2. verified that ACS can retrieve the redfish system ID and system power state
3. powering Off node works as expected, host state transition worked well and the host was indeed powered off
4. powering On also worked as expected, CloudStack host state transition as expected, host is indeed up again

### CODE REVIEW

File: utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+     */
+    protected String validateAddressAndPrepareForUrl(String hostAddress) {
+        if (NetUtils.isValidIp6(hostAddress)) {
+            return String.format("[%s]", hostAddress);
+        } else if (NetUtils.isValidIp4(hostAddress)) {
+            return hostAddress;
+        } else if (InternetDomainName.isValid(hostAddress)) {
+            return hostAddress;
+        } else {
+            throw new RedfishException(String.format("Redfish host address '%s' is not a valid IPv4 or IPv6 address", hostAddress));
```
Comment (wido) [inline utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:252]: This Exception doesn't say what the code does. A valid hostname is also accepted. Not only IPv4 or IPv6 addresses.

Comment (GabrielBrascher) [reply utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java:252]: Good point, @wido! I will update this soon

Comment (GabrielBrascher) [reply utils/src/main/java/org/apache/cloudstack/utils/redfish/RedfishClient.java]: @wido code has been updated to `Redfish host address '%s' is not a valid IPv4/IPv6 address nor a valid domain name.`

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Test LGTM, we can merge if anyone else can review and possibly share test results against a redfish BMC or simulator.

**Comment by GabrielBrascher** [comment]
@DaanHoogland you are right, docs should be updated indeed. I will work on that as well.

> and can we create a template for nested testing

Do you mean [@TestTemplate](https://junit.org/junit5/docs/5.0.2/api/org/junit/jupiter/api/TestTemplate.html) from JUnit5?

**Comment by DaanHoogland** [comment]
> Do you mean [@TestTemplate](https://junit.org/junit5/docs/5.0.2/api/org/junit/jupiter/api/TestTemplate.html) from JUnit5?

no, nice link though, @GabrielBrascher . What i mean is a template VM that can serve as hypervisor for testing in. virtual nested environment.

**Comment by wido** [comment]
@DaanHoogland The problem is that for Redfish testing (and IPMI) you need a physical machine do test this on.

**Comment by DaanHoogland** [comment]
> @DaanHoogland The problem is that for Redfish testing (and IPMI) you need a physical machine do test this on.

ok, basically means we'll have to rely on you people testing. life is a dog, possibly of female gender, or not.

**Comment by rhtyd** [comment]
I think like many other plugins, this would fall under 3rd party support due to lack of simulator/hardware. The ipmi specific plugin uses a ipmi-simulator library. I'm okay to merge it as is.

**Pull request merged**